server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -164,13 +163,43 @@ public Builder(String name, IndexVersion indexVersionCreated) {
             });
         }
 
+        public Builder(String name, IndexVersion indexVersionCreated, int dims) {
+            super(name);
+
+            final boolean indexedByDefault = indexVersionCreated.onOrAfter(INDEXED_BY_DEFAULT_INDEX_VERSION);
+            if (indexedByDefault == false) {
+                throw new UnsupportedOperationException("Dynamic dim size is not supported for index version " + indexVersionCreated);
+            }
+
+            this.indexVersionCreated = indexVersionCreated;
+
+            this.dims.setValue(dims);
+
+            this.indexed = Parameter.indexParam(m -> toType(m).indexed, true);
+            this.indexed.alwaysSerialize();
+
+            this.similarity = Parameter.enumParam(
+                "similarity",
+                false,
+                m -> toType(m).similarity,
+                (Supplier<VectorSimilarity>) () -> indexed.getValue() ? VectorSimilarity.COSINE : null,
+                VectorSimilarity.class
+            ).acceptsNull().setSerializerCheck((id, ic, v) -> v != null);
+        }
+
         @Override
         protected Parameter<?>[] getParameters() {
             return new Parameter<?>[] { elementType, dims, indexed, similarity, indexOptions, meta };
         }
 
         @Override
         public DenseVectorFieldMapper build(MapperBuilderContext context) {
+
+            // At this point we expect dims to be set, if it's not there's a problem.
+            if (dims.getValue() == null) {
+                throw new IllegalArgumentException("Missing required parameter [dims] for field [" + name + "]");
+            }
```

Comment:
I wonder if we could get away with a `null` dims and dynamically setting it via dynamic mapping updates from the parser?